internal/relationship/object_cmd.go

```diff
@@ -301,9 +225,229 @@ func (c *UnregisterObjectHandler) removeObjectRelationships(ctx context.Context,
+	}
+
+	engine, err := zanzi.NewZanzi(runtime.GetKVStore(), runtime.GetLogger())
+	if err != nil {
+		return nil, newAmendRegistrationErr(err)
+	}
+
+	policy, relRec, err := h.verifyPreconditions(ctx, engine, req)
+	if err != nil {
+		return nil, newAmendRegistrationErr(err)
+	}
+
+	_, err = engine.DeleteRelationship(ctx, policy, relRec.Relationship)
+	if err != nil {
+		return nil, newAmendRegistrationErr(errors.Wrap("removing old relationship", err))
+	}
+
+	relRec.OwnerDid = req.NewOwner.Id
+	relRec.Relationship.Subject = &types.Subject{
+		Subject: &types.Subject_Actor{
+			Actor: req.NewOwner,
+		},
+	}
+
+	_, err = engine.SetRelationship(ctx, policy, relRec)
+	if err != nil {
+		return nil, newAmendRegistrationErr(errors.Wrap("creating new relationship", err))
+	}
+
+	return &types.AmendRegistrationResponse{
+		Record: relRec,
+	}, nil
+}
+
+func (h *AmendRegistrationHandler) verifyPreconditions(ctx context.Context, engine *zanzi.Adapter, req *types.AmendRegistrationRequest) (*types.Policy, *types.RelationshipRecord, error) {
+	polRec, err := engine.GetPolicy(ctx, req.PolicyId)
+	if err != nil {
+		return nil, nil, err
+	}
+	if polRec == nil {
+		return nil, nil, errors.NewPolicyNotFound(req.PolicyId)
+	}
+
+	err = did.IsValidDID(req.NewOwner.Id)
+	if err != nil {
+		return nil, nil, errors.NewFromBaseError(err, errors.ErrorType_BAD_INPUT, "invalid actor id", errors.Pair("id", req.NewOwner.Id))
+	}
+
+	relRec, err := queryOwnerRelationship(ctx, engine, polRec.Policy, req.Object)
+	if err != nil {
+		return nil, nil, err
+	}
+	if relRec == nil {
+		return nil, nil, errors.Wrap("object not registered", errors.ErrorType_BAD_INPUT,
+			errors.Pair("policy", req.PolicyId),
+			errors.Pair("resource", req.Object.Resource),
+			errors.Pair("id", req.Object.Id),
+		)
+	}
+
+	return polRec.Policy, relRec, nil
+}
+
+type UnarchiveObjectHandler struct{}
+
+func (h *UnarchiveObjectHandler) Handle(ctx context.Context, runtime runtime.RuntimeManager, cmd *types.UnarchiveObjectRequest) (*types.UnarchiveObjectResponse, error) {
+	engine, err := zanzi.NewZanzi(runtime.GetKVStore(), runtime.GetLogger())
+	if err != nil {
+		return nil, newRegisterObjectErr(err)
+	}
+
+	rec, err := engine.GetPolicy(ctx, cmd.PolicyId)
+	if err != nil {
+		return nil, err
+	}
+	if rec == nil {
+		return nil, newRegisterObjectErr(errors.NewPolicyNotFound(cmd.PolicyId))
+	}
+	pol := rec.Policy
+
+	principal, err := auth.ExtractPrincipalWithType(ctx, auth.DID)
+	if err != nil {
+		return nil, newRegisterObjectErr(err)
+	}
+	did := principal.Identifier()
+
+	ownerRecord, err := queryOwnerRelationship(ctx, engine, pol, cmd.Object)
+	if err != nil {
+		return nil, newArchiveObjectErr(err)
+	}
+	if ownerRecord == nil {
+		return nil, newUnarchiveObjectErr(errors.Wrap("object not registered", errors.ErrorType_BAD_INPUT,
+			errors.Pair("policy", pol.Id),
+			errors.Pair("resource", cmd.Object.Resource),
+			errors.Pair("id", cmd.Object.Id),
+		))
+	}
+	if !ownerRecord.Archived {
```

Comment:
fix: if record is not archived, return err - maybe wouldn't be idempotent

Reply:
nvm this is ok, it means the system is already at the desired state and does nothing